### PR TITLE
if build resources fails, download resources from charmstore

### PR DESCRIPTION
**Pull request by addyess**

### CODE REVIEW

File: tests/integration/test_kubernetes_worker_integration.py [hunk 1]
```diff
@@ -20,37 +29,82 @@ def _check_status_messages(ops_test):
...
+        log.info("Downloading Resources...")
+        await asyncio.gather(
+            *(
+                _retrieve_url(816, arch, tmpdir / f"cni-{arch}.tgz")
+                for arch in ("amd64", "arm64", "s390x")
+            )
+        )
+        current_resources = list(Path(tmpdir).glob("*.tgz"))
+
+    yield current_resources
```
Comment (johnsca) [inline tests/integration/test_kubernetes_worker_integration.py:58]: Should it clean up resources after the fact? I usually lean toward being good stewards and cleaning up after ourselves, but it's also not useful to fail a test if cleanup fails or putting much extra time / effort into cleanup which could even make the test timeout. Also, in this case, keeping them around could also save time effort on future runs.

File: tests/integration/test_kubernetes_worker_integration.py [hunk 2]
```diff
@@ -20,37 +29,82 @@ def _check_status_messages(ops_test):
...
+
 @pytest.mark.abort_on_fail
-async def test_build_and_deploy(ops_test):
-    cni_amd64 = ops_test.tmp_path / "cni-amd64.tgz"
-    urlretrieve(CNI_AMD64_URL, cni_amd64)
-    worker_charm = await ops_test.build_charm(".")
-
-    # Work around libjuju not handling local file resources by manually
-    # pre-deploying the charm w/ resource via the CLI. See
-    # https://github.com/juju/python-libjuju/issues/223
```
Comment (johnsca) [inline tests/integration/test_kubernetes_worker_integration.py:31]: Nice, glad to see this has been resolved.

**Review by johnsca — CHANGES_REQUESTED** [review]

File: tests/integration/test_kubernetes_worker_integration.py [hunk 3]
```diff
@@ -20,37 +29,82 @@ def _check_status_messages(ops_test):
...
+        rc, stdout, stderr = await ops_test.run(
+            *shlex.split(f"sudo {build_script}"), cwd=tmpdir, check=False
+        )
+        if rc != 0:
+            log.warning(f"build-cni-resources failed: {(stderr or stdout).strip()}")
+        current_resources = list(Path(tmpdir).glob("*.tgz"))
+    if not current_resources:
+        # if we couldn't build them, just download a fixed version
+        log.info("Downloading Resources...")
+        await asyncio.gather(
```
[the remark this answers is not in the document]
Comment (addyess) [reply tests/integration/test_kubernetes_worker_integration.py:50]: `asyncio.gather` raises any exception which occurs among its tasks.  However -- i can use `return_exception` to gather up any exception and `pytest.fail(...)` in the event any exception occurs

File: tests/integration/test_kubernetes_worker_integration.py (hunk 1 above)
Comment (addyess) [reply tests/integration/test_kubernetes_worker_integration.py:58]: these are all downloaeded to a `.tox/integration/tmp` which gets deleted on the next run.  This is very similar to what `pytest_ops` does when it builds the charm or renders the bundle.  They all get cleaned on the next run.  This code doesn't attempt to reuse any other downloaded resources

Comment (johnsca) [reply tests/integration/test_kubernetes_worker_integration.py:63]: Oh, I missed that they were both using `tmpdir`.  +1

**Review by johnsca — APPROVED** [review]
LGTM

**Pull request merged**